aten/src/ATen/native/cuda/fused_adam_utils.cuh

```diff
@@ -17,20 +18,25 @@ constexpr uint8_t kExpAvgIdx = 2;
 constexpr uint8_t kExpAvgSqIdx = 3;
 constexpr uint8_t kMaxExpAvgSqIdx = 4;
 
-template <typename scalar_type, typename opmath_t, int depth = 4>
-C10_DEVICE __forceinline__ void adam_math(
+template <
+    typename scalar_type,
+    typename opmath_t,
+    int depth,
+    ADAM_MODE adam_mode,
```

Comment:
I understand that moving dynamic arguments to static templates are a good thing for performance, but does this have to do with register usage?

Reply:
could be. I wanted to be clearer about which changes mean more or less but the output of nvdisasm register lifetime was a bit too big 😓